File: android/app/src/main/java/com/sovereignnetworkmobile/web4/Web4Client.kt
package com.sovereignnetworkmobile.web4

import android.util.Log
import com.sovereignnetworkmobile.NativeQuicBridge
import org.json.JSONArray
import org.json.JSONObject

class Web4Client(
    private val baseUrl: String,
    private val timeoutSecs: Int,
    private val insecure: Boolean
) {
    private val tag = "Web4Client"

    private fun makeUrl(path: String): String {
        return if (path.startsWith("quic://") || path.startsWith("https://")) {
            path
        } else {
            baseUrl.trimEnd('/') + path
        }
    }

    fun resolveDomain(domain: String): Web4ResolveResponse {
        val payload = JSONObject().apply {
            put("domain", domain)
            put("version", JSONObject.NULL)
        }
        val url = makeUrl("/api/v1/web4/domains/resolve")
        Log.i(tag, "Resolving domain: $domain with alpn=public")

        val response = NativeQuicBridge.request(
            url = url,
            method = "POST",
            headersJson = """{"content-type":"application/json"}""",
            body = payload.toString(),
            timeoutSecs = timeoutSecs,
            insecure = insecure,
            alpn = "public"
        ) ?: throw IllegalStateException("No response from resolver")

        val status = (response["status"] as? Number)?.toInt() ?: 0
        val responseBody = response["body"] as? String ?: ""

        Log.d(tag, "Resolver response: status=$status")
        if (status !in 200..299) {
            Log.e(tag, "Resolve failed ($status): $responseBody")
            throw IllegalStateException("Resolve failed ($status): $responseBody")
        }

        if (responseBody.isEmpty()) {
            Log.e(tag, "Empty resolve response body")
            throw IllegalStateException("Empty resolve body")
        }

        val json = JSONObject(responseBody)
        Log.d(tag, "Domain resolved successfully: ${json.optString("domain")}")


        val version = if (json.has("version") && !json.isNull("version")) {
            json.optLong("version")
        } else {
            null
        }

        val resolveSpaFallback: String? = if (json.has("spa_fallback") && !json.isNull("spa_fallback")) {
            json.getString("spa_fallback")
        } else {
            null
        }

        return Web4ResolveResponse(
            domain = json.getString("domain"),
            manifest_cid = json.getString("manifest_cid"),
            version = version,
            spa = json.optBoolean("spa", false),
            spa_fallback = resolveSpaFallback
        )
    }

    fun fetchManifest(manifestCid: String): Web4Manifest {
        val payload = JSONObject().apply {
            put("cid", manifestCid)
        }
        val url = makeUrl("/api/v1/web4/content/manifest")
        Log.i(tag, "Fetching manifest: $manifestCid")

        val response = NativeQuicBridge.request(
            url = url,
            method = "POST",
            headersJson = """{"content-type":"application/json"}""",
            body = payload.toString(),
            timeoutSecs = timeoutSecs,
            insecure = insecure,
            alpn = "public"
        ) ?: throw IllegalStateException("No response from manifest server")

        val status = (response["status"] as? Number)?.toInt() ?: 0
        val responseBody = response["body"] as? String ?: ""

        Log.d(tag, "Manifest response: status=$status")
        if (status !in 200..299) {
            Log.e(tag, "Manifest fetch failed ($status): $responseBody")
            throw IllegalStateException("Manifest fetch failed ($status): $responseBody")
        }

        if (responseBody.isEmpty()) {
            Log.e(tag, "Empty manifest response body")
            throw IllegalStateException("Empty manifest body")
        }

        val json = JSONObject(responseBody)
        Log.d(tag, "Manifest fetched successfully")

        val files = mutableListOf<Web4ManifestFile>()
        val filesNode = json.opt("files")
        when (filesNode) {
            is JSONArray -> {
                for (i in 0 until filesNode.length()) {
                    val f = filesNode.getJSONObject(i)
                    files.add(
                        Web4ManifestFile(
                            path = f.getString("path"),
                            cid = f.getString("cid"),
                            mime = f.optString("mime", f.optString("content_type", "application/octet-stream")),
                            size = f.optLong("size", 0)
                        )
                    )
                }
            }
            is JSONObject -> {
                val keys = filesNode.keys()
                while (keys.hasNext()) {
                    val path = keys.next()
                    val f = filesNode.getJSONObject(path)
                    files.add(
                        Web4ManifestFile(
                            path = path,
                            cid = f.getString("cid"),
                            mime = f.optString("mime", f.optString("content_type", "application/octet-stream")),
                            size = f.optLong("size", 0)
                        )
                    )
                }
            }
            else -> {
                // No files node or unsupported shape; leave empty to allow upstream handling
            }
        }

        val version: String? = if (json.has("version") && !json.isNull("version")) {
            json.getString("version")
        } else {
            null
        }

        val domain: String? = if (json.has("domain") && !json.isNull("domain")) {
            json.getString("domain")
        } else {
            null
        }

        val previousManifest: String? = if (json.has("previous_manifest") && !json.isNull("previous_manifest")) {
            json.getString("previous_manifest")
        } else {
            null
        }

        val spaFallback: String? = if (json.has("spa_fallback") && !json.isNull("spa_fallback")) {
            json.getString("spa_fallback")
        } else {
            null
        }

        return Web4Manifest(
            domain = domain,
            version = version,
            previous_manifest = previousManifest,
            spa = json.optBoolean("spa", false),
            spa_fallback = spaFallback,
            files = files
        )
    }

    fun fetchBlob(cid: String): ByteArray {
        val payload = JSONObject().apply {
            put("cid", cid)
        }
        val url = makeUrl("/api/v1/web4/content/blob")
        Log.i(tag, "Fetching blob: $cid")

        val response = NativeQuicBridge.requestBytes(
            url = url,
            method = "POST",
            headersJson = """{"content-type":"application/json"}""",
            body = payload.toString().toByteArray(),
            timeoutSecs = timeoutSecs,
            insecure = insecure,
            alpn = "public"
        ) ?: throw IllegalStateException("No response from blob server")

        val status = (response["status"] as? Number)?.toInt() ?: 0
        val error = response["error"] as? String ?: ""

        Log.d(tag, "Blob response: status=$status, size=${(response["body"] as? ByteArray)?.size ?: 0}")
        if (status !in 200..299) {
            Log.e(tag, "Blob fetch failed ($status): $error")
            throw IllegalStateException("Blob fetch failed ($status): $error")
        }

        val body = response["body"] as? ByteArray
        if (body == null || body.isEmpty()) {
            Log.e(tag, "Empty blob response body")
            throw IllegalStateException("Empty blob body")
        }

        Log.d(tag, "Blob fetched successfully: ${body.size} bytes")
        return body
    }
}

File: src/services/Web4Client.ts
/**
 * Web4 Client - Direct QUIC calls for Web4 endpoints
 * Uses public ALPN for Web4 content endpoints (read-only, no auth)
 */

import { publicQuicRequest } from './quic';

export interface Web4ResolveResponse {
  domain: string;
  manifest_cid?: string;
  error?: string;
}

export interface Web4ManifestResponse {
  cid: string;
  files?: Record<string, unknown>;
  error?: string;
}

class Web4Client {
  /**
   * Resolve a Web4 domain to get manifest CID
   */
  async resolveDomain(domain: string): Promise<Web4ResolveResponse> {
    try {
      const data = await publicQuicRequest<Web4ResolveResponse>(
        '/api/v1/web4/domains/resolve',
        {
          method: 'POST',
          body: JSON.stringify({ domain, version: null }),
          headers: { 'content-type': 'application/json' },
        },
      );
      console.log('[Web4Client] resolveDomain:', { domain, manifest_cid: data.manifest_cid });
      return data;
    } catch (error) {
      console.error('[Web4Client] resolveDomain failed:', error);
      return { domain, error: String(error) };
    }
  }

  /**
   * Fetch Web4 manifest by CID
   */
  async fetchManifest(manifestCid: string): Promise<Web4ManifestResponse> {
    try {
      const data = await publicQuicRequest<Web4ManifestResponse>(
        '/api/v1/web4/content/manifest',
        {
          method: 'POST',
          body: JSON.stringify({ cid: manifestCid }),
          headers: { 'content-type': 'application/json' },
        },
      );
      console.log('[Web4Client] fetchManifest:', { cid: manifestCid });
      return data;
    } catch (error) {
      console.error('[Web4Client] fetchManifest failed:', error);
      return { cid: manifestCid, error: String(error) };
    }
  }

  /**
   * Fetch Web4 blob content by CID
   */
  async fetchBlob(cid: string): Promise<{ data?: string; error?: string }> {
    try {
      const data = await publicQuicRequest<{ data: string }>(
        '/api/v1/web4/content/blob',
        {
          method: 'POST',
          body: JSON.stringify({ cid }),
          headers: { 'content-type': 'application/json' },
        },
      );
      console.log('[Web4Client] fetchBlob:', { cid });
      return data;
    } catch (error) {
      console.error('[Web4Client] fetchBlob failed:', error);
      return { error: String(error) };
    }
  }
}

export default new Web4Client();
